server/service/src/stock_line/historical_stock.rs

```diff
@@ -43,24 +45,47 @@ pub fn get_historical_stock_lines(
-    let stock_movements = StockMovementRepository::new(&ctx.connection).query(Some(filter))?;
+    let mut stock_movements = StockMovementRepository::new(&ctx.connection).query(Some(filter))?;
 
-    // Calculate how much each stock line has been adjusted (unit quantity)
-    let mut stock_line_adjustments: HashMap<String, f64> = HashMap::new();
+    // sort stock movements by datetime descending (latest first)
+    stock_movements.sort_by(|a, b| b.datetime.cmp(&a.datetime));
+
+    let mut available_stock_by_line: HashMap<String, f64> = HashMap::new();
+    let mut min_available_stock_by_line: HashMap<String, f64> = HashMap::new();
+    // Calculate available stock for each stock line currently
+    for stock_line in current_stock_lines.rows.iter() {
+        let available_stock_now = stock_line.stock_line_row.available_number_of_packs
+            * stock_line.stock_line_row.pack_size;
+        available_stock_by_line.insert(stock_line.stock_line_row.id.clone(), available_stock_now);
+        min_available_stock_by_line
+            .insert(stock_line.stock_line_row.id.clone(), available_stock_now);
+    }
+
+    // Calculate min available stock for each stock line for each stock movement
     for stock_movement in stock_movements {
-        let stock_line_id = stock_movement.stock_line_id.unwrap_or_default(); // Stock line ID shouldn't be null due to the filter applied...
-        let quantity = stock_movement.quantity;
-        let adjustment = stock_line_adjustments.get(&stock_line_id).unwrap_or(&0.0) + quantity;
-        stock_line_adjustments.insert(stock_line_id, adjustment);
+        let stock_line_id = stock_movement.stock_line_id.unwrap_or_default(); // Stock line ID shouldn't be null due to the repository filter applied...
+        let available_stock =
+            available_stock_by_line.get(&stock_line_id).unwrap_or(&0.0) - stock_movement.quantity;
+
+        if available_stock
+            < *min_available_stock_by_line
+                .get(&stock_line_id)
+                .unwrap_or(&0.0)
+        {
+            min_available_stock_by_line.insert(stock_line_id.clone(), available_stock);
+        }
+        available_stock_by_line.insert(stock_line_id, available_stock);
     }
 
     // Create the historical stock lines, adjust what can be allocated at that time, based on stock available then and now
-    let mut historical_stock_lines: Vec<StockLine> = vec![];
+    let mut adjusted_stock_lines: Vec<StockLine> = vec![];
     for stock_line in current_stock_lines.rows {
         let stock_line_id = stock_line.stock_line_row.id.clone();
-        let adjustment = stock_line_adjustments.get(&stock_line_id).unwrap_or(&0.0);
-        let historical_available_packs = stock_line.stock_line_row.available_number_of_packs
-            - adjustment / stock_line.stock_line_row.pack_size;
+
+        let historical_available_packs = *min_available_stock_by_line
+            .get(&stock_line_id)
+            .unwrap_or(&0.0)
```

Comment:
We can just use the min right, no other logic needed ?

Reply:
I think that would be fine, however you will see a line for new stock in your store with 0 available.
Now that I think about it I htink that's probably good. You'll see you have the stock line but won't be able to allocated it. Hopefully it will be obvious that the date has affected the available stock?

Reply:
Thanks @andreievg ready for re-review